src/OrleansRuntime/Streams/PersistentStream/PersistentStreamPullingAgent.cs

```diff
@@ -332,80 +334,113 @@ public void RemoveSubscriber_Impl(GuidId subscriptionId, StreamId streamId)
 
-                var myQueueId = (QueueId)(state);
                 if (IsShutdown) return; // timer was already removed, last tick
                 
-                IQueueAdapterReceiver rcvr = receiver;
                 int maxCacheAddCount = queueCache?.GetMaxAddCount() ?? QueueAdapterConstants.UNLIMITED_GET_QUEUE_MSG;
 
                 // loop through the queue until it is empty.
                 while (!IsShutdown) // timer will be set to null when we are asked to shudown. 
                 {
-                    var now = DateTime.UtcNow;
-                    // Try to cleanup the pubsub cache at the cadence of 10 times in the configurable StreamInactivityPeriod.
-                    if ((now - lastTimeCleanedPubSubCache) >= config.StreamInactivityPeriod.Divide(StreamInactivityCheckFrequency))
-                    {
-                        lastTimeCleanedPubSubCache = now;
-                        CleanupPubSubCache(now);
-                    }
+                    // If read succeeds and there is more data, we continue reading.
+                    // If read succeeds and there is no more data, we breack out of loop
+                    // If read fails, we try again, with backoff policy.
+                    //    This prevents spamming backend queue which may be encountering transient errors.
+                    //    We retry until the operation succeeds or we are shutdown.
+                    bool moreData = await AsyncExecutorWithRetries.ExecuteWithRetries(
+                        i => ReadFromQueue((QueueId)state, receiver, maxCacheAddCount),
+                        AsyncExecutorWithRetries.INFINITE_RETRIES,
+                        (e, i) => !IsShutdown,
+                        TimeSpan.MaxValue,
+                        ReadLoopBackoff);
+                    if (!moreData)
+                        return;
+                }
+            }
+            catch (Exception exc)
+            {
+                receiverInitTask = null;
+                logger.Error(ErrorCode.PersistentStreamPullingAgent_12, "Exception while PersistentStreamPullingAgentGrain.AsyncTimerCallback", exc);
+            }
+        }
 
-                    if (queueCache != null)
+        /// <summary>
+        /// Read from queue.
+        /// Returns true, if data was read, false if it was not
+        /// </summary>
+        /// <param name="myQueueId"></param>
+        /// <param name="rcvr"></param>
+        /// <param name="maxCacheAddCount"></param>
+        /// <returns></returns>
+        private async Task<bool> ReadFromQueue(QueueId myQueueId, IQueueAdapterReceiver rcvr, int maxCacheAddCount)
+        {
+            try
+            {
+                var now = DateTime.UtcNow;
+                // Try to cleanup the pubsub cache at the cadence of 10 times in the configurable StreamInactivityPeriod.
+                if ((now - lastTimeCleanedPubSubCache) >= config.StreamInactivityPeriod.Divide(StreamInactivityCheckFrequency))
```

Comment:
the cache cleanup code logically does not belong inside the func, but rather outside the loop, or even better in a separate timer. you can improve it later on.